File: python/src/banking/withdraw.py
"""Lesson 1: executable preconditions and postconditions."""

from .money import Balance, WithdrawalAmount
from .verified_arithmetic import calculate_new_balance


def withdraw(balance: Balance, amount: WithdrawalAmount) -> Balance:
    """Return the balance after withdrawing ``amount``.

    Raises:
        ValueError: if the amount exceeds the balance.
        AssertionError: if this implementation violates a postcondition.
    """
    if amount.value > balance.value:
        raise ValueError("amount must not exceed balance")

    new_balance = Balance(calculate_new_balance(balance.value, amount.value))

    # These are implementation promises, not checks on the caller.
    if new_balance.value < 0:
        raise AssertionError("postcondition failed: result must be non-negative")
    if new_balance.value != balance.value - amount.value:
        raise AssertionError("postcondition failed: result must equal balance - amount")

    return new_balance

File: kotlin/src/main/kotlin/banking/Withdraw.kt
package banking

private enum class WithdrawalDecision {
    ALLOWED,
    OVERDRAFT,
}

private fun decide(balance: Balance, amount: WithdrawalAmount): WithdrawalDecision =
    if (amount.value <= balance.value) {
        WithdrawalDecision.ALLOWED
    } else {
        WithdrawalDecision.OVERDRAFT
    }

fun withdraw(balance: Balance, amount: WithdrawalAmount): Balance =
    when (decide(balance, amount)) {
        WithdrawalDecision.OVERDRAFT ->
            throw IllegalArgumentException("amount must not exceed balance")

        WithdrawalDecision.ALLOWED -> {
            val newBalance = Balance(balance.value - amount.value)

            // `check` expresses promises made by the implementation.
            check(newBalance.value >= 0) {
                "postcondition failed: result must be non-negative"
            }
            check(newBalance.value == balance.value - amount.value) {
                "postcondition failed: result must equal balance - amount"
            }

            newBalance
        }
    }
